packages/@sanity/form-builder/src/inputs/ObjectInput/ConditionalField.tsx

```diff
@@ -0,0 +1,61 @@
+import React, {useMemo} from 'react'
+import {
+  BaseSchemaType,
+  HiddenOptionCallbackContext,
+  HiddenOptionCallback,
+  SanityDocument,
+} from '@sanity/types'
+
+import withDocument from '../../utils/withDocument'
+
+type HiddenOption = BaseSchemaType['hidden']
+
+function isThenable(value: any) {
+  return typeof value?.then === 'function'
+}
+
+function checkCondition(
+  schemaHiddenOption: HiddenOptionCallback | undefined | boolean,
+  context: HiddenOptionCallbackContext
+): boolean {
+  const result =
+    typeof schemaHiddenOption === 'function' ? schemaHiddenOption(context) : schemaHiddenOption
+
+  if (isThenable(result)) {
+    console.warn(
+      '[Warning]: The hidden option is either a promise or a promise returning function. Async callbacks for `hidden` option is not currently supported.'
+    )
```

Comment:
we could store some state to prevent this warning from over logging

Reply:
Fixed in b2b4797